packages/vitest-pool-workers/src/pool/cloudflare-pool-worker.ts

```diff
@@ -0,0 +1,230 @@
+import { type WorkerPoolOptionsContext } from "./plugin";
+import {
+	assertCompatibleVitestVersion,
+	connectToMiniflareSocket,
+	getDurableObjectDesignators,
+	getProjectMiniflare,
+	getRunnerName,
+	maybeGetResolvedMainPath,
+	structuredSerializableParse,
+	structuredSerializableStringify,
+} from ".";
+import type {
+	WorkersConfigPluginAPI,
+	WorkersPoolOptions,
+	WorkersPoolOptionsWithDefines,
+} from "./config";
+import type { Miniflare, WebSocket } from "miniflare";
+import type {
+	PoolOptions,
+	PoolWorker,
+	WorkerRequest,
+	WorkerResponse,
+} from "vitest/node";
+
+export class CloudflarePoolWorker implements PoolWorker {
+	name = "cloudflare-pool";
+	private mf: Miniflare | undefined;
+	private socket: WebSocket | undefined;
+	private parsedPoolOptions: WorkersPoolOptionsWithDefines | undefined;
+	private main: string | undefined;
+
+	constructor(
+		private options: PoolOptions,
+		private poolOptions:
+			| WorkersPoolOptions
+			| ((
+					ctx: WorkerPoolOptionsContext
+			  ) => Promise<WorkersPoolOptions> | WorkersPoolOptions)
+	) {
+		assertCompatibleVitestVersion(options.project.vitest);
+	}
+
+	async start(): Promise<void> {
+		let resolvedPoolOptions: WorkersPoolOptions;
+		if (typeof this.poolOptions === "function") {
+			// hhttps://github.com/vitest-dev/vitest/blob/v4.0.15/packages/vitest/src/integrations/inject.ts
+			const inject = <K extends keyof ProvidedContext>(
+				key: K
+			): ProvidedContext[K] => {
+				return this.options.project.getProvidedContext()[key];
+			};
+			resolvedPoolOptions = await this.poolOptions({ inject });
+		} else {
+			resolvedPoolOptions = this.poolOptions;
+		}
+
+		this.parsedPoolOptions = await parseProjectOptions(
+			this.options.project,
+			resolvedPoolOptions
+		);
+		this.main = maybeGetResolvedMainPath(
+			this.options.project,
+			this.parsedPoolOptions
+		);
+
+		// Find the vitest-pool-workers plugin and give it the path to the main file.
+		// This allows that plugin to inject a virtual dependency on main so that vitest
+		// will automatically re-run tests when that gets updated, avoiding the user having
+		// to manually add such an import in their tests.
+		const configPlugin = this.options.project.vite.config.plugins.find(
+			({ name }) => name === "@cloudflare/vitest-pool-workers"
+		);
+		if (configPlugin !== undefined) {
+			const api = configPlugin.api as WorkersConfigPluginAPI;
+			api.setMain(this.main);
+		}
+
+		this.mf = await getProjectMiniflare(
+			this.options.project.vitest,
+			this.options.project,
+			this.parsedPoolOptions,
+			this.main
+		);
+
+		this.socket = await connectToMiniflareSocket(
+			this.mf,
+			getRunnerName(this.options.project)
+		);
+	}
+
+	async stop(): Promise<void> {
+		await this.mf?.dispose();
+		this.mf = undefined;
+	}
+
+	send(message: WorkerRequest): void {
+		// Vitest will always call `start()` before calling `send()`
+		assert(this.socket, "Message sent to Worker before initialisation");
```

Comment:
Is this enough to know the socket is ready? Could it be defined but not open?

Reply:
I suppose in theory? But I've never had that happen while testing, and I think the `ws.accept()` in `connectToMiniflareSocket()` should cover it.